// print-error/modules/html.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.ERROR_STACK_STYLE = undefined;
exports.default = render;
exports.renderErrorStack = renderErrorStack;

var _parse = require('./parse.js');

var _parse2 = _interopRequireDefault(_parse);

var _helpers = require('./helpers.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var default_options = {
	font_size: '20px'
};

function render(error) {
	var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

	options = (0, _helpers.merge)(default_options, (0, _helpers.convert_from_camel_case)((0, _helpers.clone)(options)));

	var result = '\n\t\t<html>\n\t\t\t<head>\n\t\t\t\t<title>Error</title>\n' + (0, _helpers.tabulate)(getErrorPageStyle(options), 4) + '\n' + (0, _helpers.tabulate)(ERROR_STACK_STYLE, 4) + '\n\t\t\t</head>\n\t\t\t<body>\n' + renderErrorStack(error) + '\n\t\t\t</body>\n\t\t</html>\n\t';

	return (0, _helpers.tabulate)(result, -2).replace(/^\n/, '').replace(/\n$/, '');
}

function renderErrorStack(error) {
	var groups = (0, _parse2.default)(error.stack);

	var groups_markup = groups.map(function (group, i) {
		var list_items_markup = group.lines.map(function (line) {
			return '<li class="print-error-stack__entry">' + '\n' + (0, _helpers.tabulate)(line_markup(line), 1) + '\n' + '</li>';
		}).join('\n');

		var markup = '\n\t\t\t<h1 class="print-error-stack__heading' + (i === 0 ? '' : ' print-error-stack__heading--secondary') + '">' + escape_html(group.title) + '</h1>\n\t\t\t<ul class="print-error-stack">\n' + (0, _helpers.tabulate)(list_items_markup, 4) + '\n\t\t\t</ul>';

		return (0, _helpers.tabulate)(markup, -3);
	}).join('').replace(/^\n/, '').replace(/\n$/, '');

	return (0, _helpers.tabulate)(groups_markup, 4);
}

function line_markup(line_info) {
	if (typeof line_info === 'string') {
		return line_info;
	}

	var line = '';

	if (line_info.file_path) {
		line += '<span class="print-error-stack__file-name">' + line_info.file_name + '</span>';
	}

	if (line_info.file_line_number) {
		line += '<span class="print-error-stack__colon">:</span><span class="print-error-stack__line-number">' + line_info.file_line_number + '</span>';
	}

	if (line_info.method_path) {
		if (line.length > 0) {
			line += ' ';
		}

		line += '<span class="print-error-stack__method">' + escape_html(line_info.method_path) + '</span>';
	}

	if (line_info.file_path) {
		line += '\n' + (0, _helpers.tabulate)('<div class="print-error-stack__file-path">\n\t\t\t' + escape_html(line_info.file_path).split('/').join('<span class="print-error-stack__file-path-separator">/</span>') + '\n\t\t</div>', -2);
	}

	return line;
}

function escape_html(text) {
	return text.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

var ERROR_STACK_STYLE = exports.ERROR_STACK_STYLE = (0, _helpers.tabulate)('\t<style>\n\t\t.print-error-stack__heading\n\t\t{\n\t\t\tfont-size : 140%;\n\t\t\tcolor     : #C44100;\n\t\t}\n\n\t\t.print-error-stack__heading--secondary\n\t\t{\n\t\t\tfont-weight : normal;\n\t\t\tcolor       : #7f7f7f;\n\t\t}\n\n\t\t.print-error-stack\n\t\t{\n\t\t\tmargin-top : 2em;\n\t\t}\n\n\t\t.print-error-stack__entry\n\t\t{\n\t\t\tmargin-bottom   : 1.5em;\n\t\t\tlist-style-type : none;\n\t\t\tfont-size       : 120%;\n\t\t}\n\n\t\t.print-error-stack__file-path\n\t\t{\n\t\t\tcolor         : #7f7f7f;\n\t\t\tmargin-top    : 0.8em;\n\t\t}\n\n\t\t.print-error-stack__file-path-separator\n\t\t{\n\t\t\tcolor : #c0c0c0;\n\t\t}\n\n\t\t.print-error-stack__file-name\n\t\t{\n\t\t\tfont-weight : bold;\n\t\t}\n\n\t\t.print-error-stack__line-number\n\t\t{\n\t\t}\n\n\t\t.print-error-stack__colon\n\t\t{\n\t\t\tcolor: #9f9f9f;\n\t\t}\n\n\t\t.print-error-stack__method\n\t\t{\n\t\t\tcolor: #0091C2;\n\t\t\tfont-weight: bold;\n\t\t}\n\t</style>', -1);

function getErrorPageStyle(options) {
	return (0, _helpers.tabulate)('\t\t<style>\n\t\t\thtml\n\t\t\t{\n\t\t\t\tfont-family : Monospace, Arial;\n\t\t\t\tfont-size   : ' + options.font_size + ';\n\t\t\t}\n\n\t\t\tbody\n\t\t\t{\n\t\t\t\tmargin-top    : 1.6em;\n\t\t\t\tmargin-bottom : 1.6em;\n\t\t\t\tmargin-left   : 2.3em;\n\t\t\t\tmargin-right  : 2.3em;\n\t\t\t}\n\t\t</style>', -2);
}
