// @catamphetamine/found/lib/cjs/Link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _useEventCallback = _interopRequireDefault(require("@restart/hooks/useEventCallback"));

var _react = _interopRequireDefault(require("react"));

var _tinyWarning = _interopRequireDefault(require("tiny-warning"));

var _useRouter = _interopRequireDefault(require("./useRouter"));

var _jsxRuntime = require("react/jsx-runtime");

const _excluded = ["as", "to", "activeClassName", "activeStyle", "activePropName", "match", "router", "exact", "onClick", "target"];

function Link(_ref) {
  let {
    as: Component = 'a',
    to,
    activeClassName,
    activeStyle,
    activePropName,
    match: propsMatch,
    router: propsRouter,
    exact = false,
    onClick,
    target
  } = _ref,
      props = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  const {
    router,
    match
  } = (0, _useRouter.default)() || {
    match: propsMatch,
    router: propsRouter
  };
  const handleClick = (0, _useEventCallback.default)(event => {
    if (onClick) {
      onClick(event);
    } // Don't do anything if the user's onClick handler prevented default.
    // Otherwise, let the browser handle the link with the computed href if the
    // event wasn't an unmodified left click, or if the link has a target other
    // than _self.


    if (event.defaultPrevented || event.metaKey || event.altKey || event.ctrlKey || event.shiftKey || event.button !== 0 || target && target !== '_self') {
      return;
    }

    event.preventDefault(); // FIXME: When clicking on a link to the same location in the browser, the
    // actual becomes a replace rather than a push. We may want the same
    // handling – perhaps implemented in the Farce protocol.

    router.push(to);
  });

  if (process.env.NODE_ENV !== "production" && typeof Component !== 'function') {
    for (const wrongPropName of ['component', 'Component']) {
      const wrongPropValue = props[wrongPropName];

      if (!wrongPropValue) {
        continue;
      }

      process.env.NODE_ENV !== "production" ? (0, _tinyWarning.default)(false, 'Link to %s with `%s` prop `%s` has an element type that is not a component. The expected prop for the link component is `as`.', JSON.stringify(to), wrongPropName, wrongPropValue.displayName || wrongPropValue.name || 'UNKNOWN') : void 0;
    }
  }

  const href = router.createHref(to);
  const childrenIsFunction = typeof props.children === 'function';

  if (childrenIsFunction || activeClassName || activeStyle || activePropName) {
    const toLocation = router.createLocation(to);
    const active = router.isActive(match, toLocation, {
      exact
    });

    if (childrenIsFunction) {
      return props.children({
        href,
        active,
        onClick: handleClick
      });
    }

    if (active) {
      if (activeClassName) {
        props.className = props.className ? `${props.className} ${activeClassName}` : activeClassName;
      }

      if (activeStyle) {
        props.style = Object.assign({}, props.style, activeStyle);
      }
    }

    if (activePropName) {
      props[activePropName] = active;
    }
  }

  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, Object.assign({}, props, {
    href: href,
    onClick: handleClick // This overrides props.onClick.

  }));
}

var _default = Link;
exports.default = _default;
module.exports = exports.default;
