// react-pages/commonjs/components/Loading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loading = Loading;
exports["default"] = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactRedux = require("react-redux");
var _classnames = _interopRequireDefault(require("classnames"));
var _LoadingIndicator = _interopRequireDefault(require("./LoadingIndicator.js"));
var _FadeInOut = _interopRequireDefault(require("./FadeInOut.js"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function Loading(_ref) {
  var initial = _ref.initial,
    pending = _ref.pending,
    immediate = _ref.immediate,
    Indicator = _ref.indicator,
    fadeOutDuration = _ref.fadeOutDuration;
  _react["default"].createElement('div', {
    className: (0, _classnames["default"])('rrui__fixed-full-width', 'react-pages__loading', {
      'react-pages__loading--initial': initial,
      'react-pages__loading--shown': pending,
      'react-pages__loading--immediate': immediate
    })
  }, _react["default"].createElement(_FadeInOut["default"], {
    show: pending,
    fadeOutDuration: fadeOutDuration
  }, _react["default"].createElement(Indicator, {
    className: 'react-pages__loading-spinner'
  })));
}
Loading.propTypes = {
  initial: _propTypes["default"].bool,
  pending: _propTypes["default"].bool,
  immediate: _propTypes["default"].bool,
  indicator: _propTypes["default"].elementType.isRequired,
  fadeOutDuration: _propTypes["default"].number.isRequired
};
Loading.defaultProps = {
  indicator: _LoadingIndicator["default"],
  fadeOutDuration: 160
};
var _default = (0, _reactRedux.connect)(function (_ref2) {
  var preload = _ref2.preload;
  return {
    initial: preload.initial,
    pending: preload.pending,
    immediate: preload.immediate
  };
})(Loading);
exports["default"] = _default;
