// react-pages/commonjs/redux/preload/chain.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.chain_preloaders = chain_preloaders;
exports["default"] = generatePreloadChain;
exports.filter_preloaders = filter_preloaders;
exports.promisify = promisify;
var _flags = require("../../client/flags.js");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
// Returns function returning a Promise
// which resolves when all the required page `load`s are resolved.
//
// If no preloading is needed, then returns nothing.
//
function generatePreloadChain(preloaders, server, isInitialClientSidePreload, getState, dispatch, location, params, history, getCookie, preloading) {
  // Set `.preload({ ... })` function arguments,
  // and also set default `load` options.
  setUpPreloaders(preloaders, {
    dispatch: dispatch,
    getState: getState,
    location: location,
    params: params,
    history: history,
    server: server,
    // `getCookie` `load` parameter has been requested:
    // https://github.com/catamphetamine/react-website/issues/71
    getCookie: getCookie
  }, server);

  // Only select those `load`s which
  // should be run in current circumstances.
  preloaders = filter_preloaders(preloaders, server, isInitialClientSidePreload);

  // Construct a sequential chain out of preloaders.
  // Because each of them could be either parallel or sequential.
  var chain = chain_preloaders(preloaders);

  // If there are no `load`s for this route, then exit.
  if (chain.length === 0) {
    return;
  }

  // Return a function which generates preloading `Promise` chain.
  return function () {
    return promisify(chain, preloading);
  };
}

// Applies default `load` options
// and sets `load` function arguments.
function setUpPreloaders(preloaders, preload_arguments, server) {
  var _iterator = _createForOfIteratorHelper(preloaders),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var component_preloaders = _step.value;
      var _iterator2 = _createForOfIteratorHelper(component_preloaders),
        _step2;
      try {
        for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
          var preloader = _step2.value;
          setUpPreloader(preloader, preload_arguments, server);
        }
      } catch (err) {
        _iterator2.e(err);
      } finally {
        _iterator2.f();
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
}

// Applies default `load` options
// and sets `load` function arguments.
function setUpPreloader(preloader, preload_arguments, server) {
  var preload = preloader.preload;
  preloader.preload = function () {
    return preload(preload_arguments);
  };

  // If Server-Side Rendering is not being used at all
  // then all `load`s must be marked as client-side ones.
  if (!server && !(0, _flags.isServerSidePreloaded)()) {
    preloader.options.client = true;
  }
}

// Selects only those `load`s which
// should be run in current circumstances.
function filter_preloaders(preloaders, server, isInitialClientSidePreload) {
  // `preloaders` array will be mutated
  preloaders = preloaders.slice();
  preloaders.forEach(function (_, i) {
    preloaders[i] = preloaders[i].filter(function (preloader) {
      // Don't execute client-side-only `load`s on server side
      if (preloader.options.client && server) {
        return false;
      }
      // Don't execute server-side-only `load`s on client side
      if (preloader.options.server && !server) {
        return false;
      }
      // If it's initial client side preload (after the page has been loaded),
      // then only execute those `load`s marked as "client-side-only".
      if (isInitialClientSidePreload && !preloader.options.client) {
        return false;
      }
      return true;
    });
  });
  return preloaders.filter(function (component_preloaders) {
    return component_preloaders.length > 0;
  });
}

// Constructs `preload` chain.
//
// @param `preloaders` is an array of `component_preloaders`.
// `component_preloaders` is an array of all
// `load`s for a particular React component.
// Therefore `preloaders` is an array of arrays.
function chain_preloaders(preloaders) {
  var chain = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var parallel = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  // If all `preload`s have been visited
  if (preloaders.length === 0) {
    if (parallel.length === 0) {
      return chain;
    }
    // Finalize pending parallel `preload`s
    if (parallel.length === 1) {
      return chain.concat(parallel);
    }
    return chain.concat({
      parallel: parallel
    });
  }

  // `component_preloaders` is an array of all
  // `load`s for a particular React component.
  var preloader = preloaders[0];
  preloaders = preloaders.slice(1);
  if (!is_preloader_blocking(preloader)) {
    return chain_preloaders(preloaders, chain, concat(parallel, get_preloader(preloader)));
  }
  if (parallel.length === 0) {
    return chain_preloaders(preloaders, concat(chain, get_preloader(preloader)), []);
  }
  return chain_preloaders(preloaders, chain.concat({
    parallel: concat(parallel, get_preloader(preloader))
  }), []);
}
function get_preloader(preloader) {
  // A list of same component's `load`s
  if (Array.isArray(preloader)) {
    return chain_preloaders(preloader);
  }

  // Same component adjacent `load`
  return preloader.preload;
}
function is_preloader_blocking(preloader) {
  // A list of same component's `load`s
  if (Array.isArray(preloader)) {
    // Determine the proper `blocking` option
    // for this component's `load`s.
    var _iterator3 = _createForOfIteratorHelper(preloader),
      _step3;
    try {
      for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
        var sibling_preloader = _step3.value;
        // If any of component's `load`s are `blocking: true`
        // then all of them are `blocking: true`.
        if (sibling_preloader.options.blocking) {
          return true;
        }
      }
    } catch (err) {
      _iterator3.e(err);
    } finally {
      _iterator3.f();
    }
    return false;
  }

  // Same component adjacent `load`
  return preloader.options.blockingSibling;
}

// Returns a `Promise` chain.
function promisify(chain, preloading) {
  if (typeof chain === 'function') {
    return chain();
  }
  if (_typeof(chain) === 'object' && chain.parallel) {
    return Promise.all(chain.parallel.map(function (link) {
      return promisify(link, preloading);
    })).then(function (results) {
      return results.reduce(function (combined, result) {
        return _objectSpread(_objectSpread({}, combined), result);
      }, {});
    });
  }
  return chain.reduce(function (promise, link) {
    return promise.then(function (result) {
      if (preloading.cancelled) {
        return;
      }
      return promisify(link, preloading).then(function (nextResult) {
        return _objectSpread(_objectSpread({}, result), nextResult);
      });
    });
  }, Promise.resolve());
}
function concat(array, part) {
  if (Array.isArray(part) && part.length > 1) {
    // Pushes an array
    return array.concat([part]);
  }
  // Pushes a single element
  return array.concat(part);
}
